### PR TITLE
MCMC plot examples

### PR DESCRIPTION
**Pull request by chonlei**
4 new MCMC plots with a clean history.

see e.g. #63 #66

### CODE REVIEW

**Comment by MichaelClerx** [comment]
Autocorrelation: https://github.com/pints-team/pints/blob/clean-mcmc-plot-examples/examples/plot-mcmc-autocorrelation-plot.ipynb

**Comment by MichaelClerx** [comment]
Pairwise scatter: https://github.com/pints-team/pints/blob/clean-mcmc-plot-examples/examples/plot-mcmc-pairwise-scatterplots.ipynb

**Comment by MichaelClerx** [comment]
Predictions: https://github.com/pints-team/pints/blob/clean-mcmc-plot-examples/examples/plot-mcmc-predicted-time-series.ipynb

**Comment by MichaelClerx** [comment]
Trace plots: https://github.com/pints-team/pints/blob/clean-mcmc-plot-examples/examples/plot-mcmc-trace-plots.ipynb

**Comment by ben18785** [comment]
Nice work Chon. Looking good to me. A few things:

- Pairwise plots: I think I like the KDE matrix scatterplots versions. I'm not on my computer now and so can't run them now. Do they take long to run?
- Autocorrelation: worth having, although only plot the first 10 or so lags (not 350 lags).
- Predictions: I think the figure legend needs to have the 'inferred series' line without opacity (the plot should still plot with opacity).
- Trace plots: look fine. I personally don't find trace plots that useful, although I know other people do.

@MichaelClerx An aside, but can we perhaps rename the 'BayesianLogLikelihood' function in PINTS? I think it is confusing as people don't really use this terminology -- a log likelihood is not inherently Bayesian although I know you mean this as the likelihood * prior. How about 'LogPosterior'?

@MichaelClerx Another aside, sorry. Do the optimisation examples work off of data with noise added? If not, I would suggest this is worth adding. Also, can we have an actual versus fitted at the end of these examples? Think it would make the case better than just printing the numbers...(thinking https://github.com/pints-team/pints/blob/master/examples/optimisation-xnes.ipynb, https://github.com/pints-team/pints/blob/master/examples/optimisation-snes.ipynb, https://github.com/pints-team/pints/blob/master/examples/optimisation-cmaes.ipynb).

**Comment by sanmitraghosh** [comment]
I think "LogPosterior" is the correct name for the Bayesian inference

**Comment by martinjrobins** [comment]
These plots look great, thanks @chonlei. I would say that some of the explanation and plot labels are a bit sparse in some of them. For example, in the pairwise scatter plot, what is a kde plot? and some plots don't have x-axis labels, or legends indicating, for example, that the green lines indicate the true value (I know it might be obvious, but its nice to be able to skim down an example and get all the information you need from the figures)

And also the print statements seem a bit out of place for what is a static notebook, I'd probably get rid of them. Even for a dynamic notebook, I guess you already get a visual indication that a cell has finished running?

**Comment by MichaelClerx** [comment]
It would be great if @ben18785 and @sanmitraghosh could update the explanatory text in these notebook!

**Comment by MichaelClerx** [comment]
I might split the pairwise example into two notebooks: one with and one without KDE. Thoughts?

**Comment by chonlei** [comment]
Thank you all!

@ben18785 

* Pairwise plots: I think I like the KDE matrix scatterplots versions. I'm not on my computer now and so can't run them now. Do they take long to run?
   - Surprisingly fast using scipy KDE to do the plot.

@martinjrobins 

Good point! Yes, you are right, I kind of stopped quite abruptly the other day, so this has the main plots in it, but not yet "polished" for the script. I think @MichaelClerx was doing that(?), if not, I can add them in later.

---
Also, shall we add all those plots as part of the Pints and change those example by just using our new functions? 

(I am very tempted to do that - especially for the matrix plot, if I were the user, I definitely would not want to stick that "long-and-annoying two for-loops" everywhere in my script.) 

Any thoughts? I think it wouldn't be too much extra work?

**Comment by ben18785** [comment]
@Chon -- is surprisingly fast still fast for 100,000 samples? Also, I agree
with you about functions -- I thought this was the intention?

**Comment by chonlei** [comment]
Just did a quick test on how fast is the KDE from Scipy.

For 100,000 samples (excluded all thinning and burn-in), remains very fast for 1 variate, but it is no more surprising for multivariate plots. For a single multivariate plot: 
KDE: 38.082449913 s
histogram: 0.0163419246674 s
(Can't do much, coming from Scipy, I doubt it would be easy to beat it.) @ben18785

File: examples/plot-mcmc-autocorrelation.ipynb [hunk 1]
```diff
@@ -108,7 +117,9 @@
    "source": [
     "pl.figure()\n",
     "pl.acorr(chain[:,0] - np.mean(chain[:,0]))\n",
-    "pl.xlim(0, 10)\n",
+    "pl.xlim(-0.5, 10.5)\n",
+    "pl.xlabel('Lag')\n",
+    "pl.ylabel('Autocorrelation')\n",
```
Comment (MichaelClerx) [inline examples/plot-mcmc-autocorrelation.ipynb]: Thanks Chon!

**Comment by MichaelClerx** [comment]
This is merged now (So everyone can delete their local branch).

@ben18785 If you feel like elaborating on the descriptions of the plots (the bit of text near the bottom) that's still very welcome :-)

Nice work @chon!

**Pull request merged**